core/src/params.rs

```diff
@@ -0,0 +1,248 @@
+// shall be included in all copies or substantial portions
+// of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
+// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! RPC parameters.
+
+use crate::traits::ToRpcParams;
+use crate::Error;
+use serde::Serialize;
+use serde_json::value::RawValue;
+
+/// Helper module for building parameters.
+mod params_builder {
+	use serde::Serialize;
+
+	/// Initial number of bytes for a parameter length.
+	const PARAM_BYTES_CAPACITY: usize = 128;
+
+	/// Generic parameter builder that serializes parameters to bytes.
+	/// This produces a JSON compatible String.
+	///
+	/// The implementation relies on `Vec<u8>` to hold the serialized
+	/// parameters in memory for the following reasons:
+	///   1. Other serialization methods than `serde_json::to_writer` would internally
+	///      have an extra heap allocation for temporarily holding the value in memory.
+	///   2. `io::Write` is not implemented for `String` required for serialization.
+	#[derive(Debug)]
+	pub(crate) struct ParamsBuilder {
+		bytes: Vec<u8>,
+		end: char,
+	}
+
+	impl ParamsBuilder {
+		/// Construct a new [`ParamsBuilder`] with custom start and end tokens.
+		/// The inserted values are wrapped by the _start_ and _end_ tokens.
+		fn new(start: char, end: char) -> Self {
+			let mut bytes = Vec::with_capacity(PARAM_BYTES_CAPACITY);
+			bytes.push(start as u8);
+			ParamsBuilder { bytes, end }
+		}
+
+		/// Construct a new [`ParamsBuilder`] for positional parameters equivalent to a JSON array object.
+		pub(crate) fn positional() -> Self {
+			Self::new('[', ']')
+		}
+
+		/// Construct a new [`ParamsBuilder`] for named parameters equivalent to a JSON map object.
+		pub(crate) fn named() -> Self {
+			Self::new('{', '}')
+		}
+
+		/// Insert a named value (key, value) pair into the builder.
+		/// The _name_ and _value_ are delimited by the `:` token.
+		pub(crate) fn insert_named<P: Serialize>(&mut self, name: &str, value: P) -> Result<(), serde_json::Error> {
+			serde_json::to_writer(&mut self.bytes, name)?;
+			self.bytes.push(b':');
+			serde_json::to_writer(&mut self.bytes, &value)?;
+			self.bytes.push(b',');
+
+			Ok(())
+		}
+
+		/// Insert a plain value into the builder.
+		pub(crate) fn insert<P: Serialize>(&mut self, value: P) -> Result<(), serde_json::Error> {
+			serde_json::to_writer(&mut self.bytes, &value)?;
+			self.bytes.push(b',');
+
+			Ok(())
+		}
+
+		/// Finish the building process and return a JSON compatible string.
+		pub(crate) fn build(mut self) -> String {
```

Comment:
You know this already but just to write it down:

Return `Option<String>` from this, so that it can be zero alloc if no params are inserted.
Then `ObjectParams::new()` and `ArrayParams::new()` also become zero alloc and the `to_rpc_params` macro can just spit out `ArrayParams` always and be zero alloc if empty :)